File: erlang/grains/src/grains.erl
-module(grains).

-export([square/1, total/0]).

square(Square) when (Square =< 64) and (Square >= 1) ->
    trunc(math:pow(2, Square - 1));
square(_) ->
    {error, "square must be between 1 and 64"}.

total() ->
    lists:sum(
      lists:map(fun square/1, lists:seq(1, 64))).

File: elixir/grains/lib/grains.ex
defmodule Grains do
  @doc """
  Calculate two to the power of the input minus one.
  """
  @spec square(pos_integer) :: pos_integer
  def square(number) when number >= 1 and number <= 64 do
    {:ok, :math.pow(2,number - 1) |> round() }
  end
  def square(number) do
    {:error, "The requested square must be between 1 and 64 (inclusive)"}
  end

  @doc """
  Adds square of each number from 1 to 64.
  """
  @spec total :: pos_integer
  def total do
    {:ok, 1..64 |> Stream.map(fn i ->
      {:ok, res } = square(i)
      res
    end) |> Enum.sum() }
  end
end
